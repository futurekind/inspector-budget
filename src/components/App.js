import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { withRouter } from 'react-router';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group'

import * as applicationSelectors from '../redux/selectors/application';
import { saveStateToServer } from '../redux/actions/application'
import 'normalize.css/normalize.css';
import Navigation from './common/Navigation';
import NavigationBtn from './common/NavigationBtn';
import Footer from './common/Footer';

const View = styled.div`
    width: 100%;
    height: calc(100vh - 20px);
    display: flex;
    flex-wrap: wrap;
`

const Aside = styled.aside`
    width: 90px;
    height: 100%;
`

const Main = styled.main`
    height: 100%
    display: flex;
    flex: 1;
`

const TransitionHelper = styled.div`
    flex: 1;
`

export class App extends Component {

    constructor(props) {
        super(props);

        this.handleNavigationItemClick = this.handleNavigationItemClick.bind(this);
        this.handleSave = this.handleSave.bind(this);
    }

    componentDidMount () {
        document.addEventListener('keydown', this.handleSave)
    }
    
    componentWillUnmount () {
        document.removeEventListener('keydown', this.handleSave)
    }
    
    render () {
        const { activeNavigationIndex, children, location, dirty, lastSave } = this.props;
        
        return (
            <View>
                <Aside>
                    <Navigation
                        activeIndex={ activeNavigationIndex }
                        onItemClick={ this.handleNavigationItemClick }
                    >
                        <NavigationBtn
                            to="/accounts" 
                            icon="store" 
                            active={ activeNavigationIndex === 0 }
                        >Accounts</NavigationBtn>
                        <NavigationBtn
                            to="/budgets" 
                            icon="store" 
                            active={ activeNavigationIndex === 1 }
                        >Budgets</NavigationBtn>
                    </Navigation>
                </Aside>
                <Main>
                    <ReactCSSTransitionGroup
                        component={ TransitionHelper } 
                        transitionName="page"
                        transitionEnterTimeout={500} 
                        transitionLeaveTimeout={500}
                    >
                        { React.cloneElement(children, { key: location.pathname }) }
                    </ReactCSSTransitionGroup>
                    
                </Main>
                <Footer
                    items={[
                        `Last save: ${lastSave || '-'}`,
                        dirty > 0 
                            ? `${dirty} Updates`
                            : 'up to date'
                    ]}
                />
            </View>
        )
    }

    handleNavigationItemClick(index, props) {
        const { to } = props;
        const { router } = this.props;

        router.push(to);
    }

    handleSave(e) {
        const { metaKey, keyCode } = e;
        const { dispatch } = this.props;

        if(metaKey && keyCode === 83) {
            e.preventDefault();
            dispatch(saveStateToServer())
        }
    }

}

const mapState = state => {
    return {
        activeNavigationIndex: applicationSelectors.getNavIndex(state),
        dirty: applicationSelectors.getDirty(state),
        lastSave: applicationSelectors.getLastSave(state)
    }
}

export default withRouter(connect(mapState)(App))