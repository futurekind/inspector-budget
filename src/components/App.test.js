import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { App } from './App';

describe('App', () => {
    const wrapper = shallow(
        <App location={{}} activeNavigationIndex={-1}>
            <h1>Hallo</h1>
        </App>
    );

    it('renders correctly', () => {
        expect(toJSON(wrapper)).toMatchSnapshot()
    })


})