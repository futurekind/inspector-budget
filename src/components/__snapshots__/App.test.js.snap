exports[`App renders correctly 1`] = `
<styled.div>
  <styled.aside>
    <Navigation
      activeIndex={-1}
      onItemClick={[Function]}>
      <NavigationBtn
        active={false}
        icon="store"
        to="/accounts">
        Accounts
      </NavigationBtn>
      <NavigationBtn
        active={false}
        icon="store"
        to="/budgets">
        Budgets
      </NavigationBtn>
    </Navigation>
  </styled.aside>
  <styled.main>
    <ReactCSSTransitionGroup
      component={[Function]}
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="page">
      <h1>
        Hallo
      </h1>
    </ReactCSSTransitionGroup>
  </styled.main>
</styled.div>
`;
